preprocessing/analyze_chars.py

```diff
@@ -1,80 +1,88 @@
-# Use a screenshot of all ASCII chars to determine the luminance value of each of them
+"""Use a screenshot of all ASCII chars to determine the
+luminance value of each of them."""
 
 import os
 import numpy as np
 import json
 from math import floor
 from PIL import Image
+from functools import partial
+
+
+# Size of each characters in the screenshot
+CHAR_WIDTH = 11
+CHAR_HEIGHT = 23
+
+def closest_lookup(dictionary, key):
+	"""Will find the value for the key which is closest to the key given.
+
+	The dictionary keys should all be numbers."""
+	return dictionary[min(dictionary.keys(), key=lambda x:abs(x - key))]
+
+
+def coordinates_of_char(char_code):
+	"""Given a char_code, will return the coordinates of the char in chars.png"""
+	i = char_code - 33
+	x = (i % 8) * CHAR_WIDTH
+	y = (i // 8) * CHAR_HEIGHT
+	return x, y
 
 def average_val(pixels, y, x, y_step, x_step):
-    # Return average value in boolean pixels numpy array in the given rectangle
-    # Expects rectangle coordinates to be in the range of the pixels array
+    """Return average value in boolean pixels numpy array in the given rectangle
+    Expects rectangle coordinates to be in the range of the pixels array."""
     if y_step == 0:
         raise Exception('y_step must be non-zero')
     if x_step == 0:
         raise Exception('x_step must be non-zero')
 
-    start_y = int(floor(y))
-    end_y = int(floor(y+y_step))
-    start_x = int(floor(x))
-    end_x= int(floor(x+x_step))
+    end_y = y + y_step
+    end_x = x + x_step
 
-    total = 0
-    count = 0
+	dark_pixels = pixels[y:end_y, x:end_x].sum()
+    all_pixels = x_step * y_step
 
-    for py in range(start_y, end_y+1):
-        for px in range(start_x, end_x+1):
-            if pixels[py, px]:
-                total += 1
-            count += 1
+    return dark_pixels / all_pixels
 
-    return total / count
+def main():
+	"""Use a screenshot of all ASCII chars to determine the
+	luminance value of each of them."""
+	img = Image.open('chars.png').convert('1')
+	pixels = np.asarray(img)
 
-img = Image.open('chars.png').convert('1')
-pixels = np.asarray(img)
+	# Image contains character codes 33-127 inclusive, in 12 rows and 8 columns.
+	# The last entry is blank
+	# We'll wrap that around to use as the space char (char code 32).
 
-# Size of each characters in the screenshot
-CHAR_WIDTH = 10
```

Comment:
I think you were getting off by one errors from this -- everywhere you have been adding one, when really this should be higher by 1

Reply:
wat